### DISCUSSION TITLE
Using default for datetime query parameter without timezone leads to OpenAPI SpecValidationException

### DISCUSSION THREAD

**Discussion by johnthagen** [fastapi/fastapi#8981] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from datetime import datetime

from fastapi import FastAPI
import uvicorn

app = FastAPI()

start = datetime.today()


@app.get("/")
async def root(start_date: datetime = start):
    print(start_date)
    return {"start_date": start_date}

if __name__ == "__main__":
    uvicorn.run(app, host="127.0.0.1", port=8000)
```

### Description

Run [`openapi-generator`](https://github.com/OpenAPITools/openapi-generator) and see the following error:

```bash
$ openapi-generator --version
openapi-generator-cli 4.3.1
  commit : 003165c
  built  : 2020-05-06T09:38:39Z
  source : https://github.com/openapitools/openapi-generator
  docs   : https://openapi-generator.tech/

$ openapi-generator generate --input-spec http://localhost:8000/openapi.json --output api --generator-name typescript-axios
Exception in thread "main" org.openapitools.codegen.SpecValidationException: There were issues with the specification. The option can be disabled via validateSpec (Maven/Gradle) or --skip-validate-spec (CLI).
 | Error count: 1, Warning count: 0
Errors: 
	-attribute paths.'/'(get).parameters.[start_date].schemas.default=`2020-12-15T07:57:25.277303` is not of type `date-time`

	at org.openapitools.codegen.config.CodegenConfigurator.toContext(CodegenConfigurator.java:480)
	at org.openapitools.codegen.config.CodegenConfigurator.toClientOptInput(CodegenConfigurator.java:507)
	at org.openapitools.codegen.cmd.Generate.execute(Generate.java:423)
	at org.openapitools.codegen.cmd.OpenApiGeneratorCommand.run(OpenApiGeneratorCommand.java:32)
	at org.openapitools.codegen.OpenAPIGenerator.main(OpenAPIGenerator.java:61)
```

I would expect FastAPI to handle default `datetime` objects in a way that is compatible with standard OpenAPI generation tools.

- `datetime` is listed as a supported extra data type: https://fastapi.tiangolo.com/tutorial/extra-data-types/#other-data-types

### Environment

* OS: macOS 10.15.7
* FastAPI Version: 0.62
* Python version: 3.8.6

**Comment by johnthagen** [comment]
This seems to be a UTC issue. Should FastAPI warn or auto-convert to UTC?

The following code worked:

```py
from datetime import datetime

from fastapi import FastAPI
import uvicorn

app = FastAPI()

# Must use UTC datetime.
start = datetime.utcnow()


@app.get("/")
async def root(start_date: datetime = start):
    print(start_date)
    return {"start_date": start_date}

if __name__ == "__main__":
    uvicorn.run(app, host="127.0.0.1", port=8000)
```

**Comment by Mause** [comment]
Try

```python
async def root(start_date: str = Query(str(start))):
    ...
```
instead

**Comment by johnthagen** [comment, marked as answer]
[According to the Python docs](https://docs.python.org/3.8/library/datetime.html#datetime.datetime.utcnow), a better way to get a timezone aware `datetime` is:

```py
from datetime import datetime, timezone

from fastapi import FastAPI
import uvicorn

app = FastAPI()

start = datetime.now(timezone.utc)


@app.get("/")
async def root(start_date: datetime = start):
    print(start_date)
    return {"start_date": start_date}

if __name__ == "__main__":
    uvicorn.run(app, host="127.0.0.1", port=8000)
```

**Comment by johnthagen** [comment]
@Mause Thank you for the suggestion. The issue is that I would like the API to be as descriptive as possible, so I would like the `date-time` type to be propagated into the generated API:

<img width="632" alt="Screen Shot 2020-12-15 at 8 36 39 AM" src="https://user-images.githubusercontent.com/10340167/102221864-ae0a2a00-3eb0-11eb-8617-5b383b8fcec8.png">

This works in the examples above if I use UTC timezone explicitly.

If this is a known issue, it would be good to add this to the docs so users are aware. I'd be happy to open a PR if someone on the team is able to confirm that this is something that users need to be aware of.

**Comment by Mause** [comment]
Then try
```python
async def root(start_date: datetime = Query(datetime.now(timezone.utc).isoformat())):
    ...
```
instead?

**Comment by johnthagen** [comment]
The `isoformat()` returns a `str` so it is not needed. FastAPI converts to ISO format correctly by default. I think the issue is that if timezone information is not available, FastAPI still just serializes the `datetime`, but it is not a properly formed OpenAPI `date-time` (if missing the timezone).

A simplification of my earlier working example is:

```py
async def root(start_date: datetime = datetime.now(timezone.utc)):
    ...
```

But I before closing this issue, I think a documentation note would be helpful, and perhaps maintainers of FastAPI want to decide if passing non-timezone `datetime`s should raise a warning or error given they don't conform to the Spec they generate.

**Comment by Mause** [comment]
Feel free to raise an issue with the library used for schema generation - https://github.com/samuelcolvin/pydantic

**Comment by Mause** [comment]
The documentation for the date formats used is here: https://pydantic-docs.helpmanual.io/usage/types/#datetime-types

**Comment by johnthagen** [comment]
Another reference, it does seem to be that JSON Schema requires a time zone to be a valid `date-time`: https://github.com/OAI/OpenAPI-Specification/issues/1498#issuecomment-369680369
